Share the VM and native lambda AST in a single type.

ppedrot opened the pull request:
This PR is the "when they lead me out the door, I come back through the window" answer to #16629. In this PR, we try to share as much infrastructure as possible in the first phase of the VM / native pipeline, known as the `lambda` type. In particular, the two AST types have been merged into a single one, removing discrepancies that accumulated.

This is not yet sharing the full translation phase from Constr.t to lambda, but this is a good step towards it. For instance, many utility functions have been shared, reducing the divergence creeping up into the two implementations.

ppedrot commented:
@coqbot bench @coqbot run full ci

ppedrot commented:
It seems coqbot doesn't like receive two orders at once, so @coqbot bench

Alizter commented:
Bench looks like just noise. Can we have a bench with coq-native?

ppedrot commented:
I have tweaked the bench script manually, as usual. @coqbot bench

JasonGross commented:
What happened to coq-fiat-crypto-with-bedrock?  (The bench logs don't seem to say)

JasonGross commented:
Oh, I, see
```
# /bin/sh: 17: ulimit: error setting limit (Invalid argument)
```
Does ulimit -S -s not work on the bench?

JasonGross commented:
Once https://github.com/mit-plv/fiat-crypto/pull/1445 is merged (in about 6 hours), the bench should work with fiat-crypto again.  Why is the hard stack limit on the bench set so small, though?  (I'm presuming that's the issue.)

SkySkimmer commented:
Last we touched bench ulimit was https://github.com/coq/coq/commit/f7413f89c6dae529a4a828c6fba35e745c13c710

ppedrot commented:
When tweaking the representation of Lproj native nodes, I introduced a small discrepancy that might explain the observed speed-ups in metacoq. They were treated as always substitutable by the simplification process because they were always values, but since the record is now part of the node, this may trigger a code blowup if the argument is big. In the case of metacoq this is probably beneficial but just to err on the safe side I have added a commit to restore the previous behaviour.

Let's see what this gives. @coqbot bench native

silene commented:
Overall, this looks very good. But could you please modify the original post with the discrepancies that you found (and I presume, no longer exist) to ease the review process? For example, where does this `Lmeta` thing come from and why should we care?

ppedrot commented:
@silene some discrepancies were removed, but not all. The commits are atomic w.r.t. such differences, so I guess you can just look at the title of each commit to see what happened.

ppedrot commented:
For `Lmeta` it's just that native is able to handle normalization of metavariables, but not the VM. So the latter had no node in lambda for this. Now the VM still doesn't handle metas, but it fails one level below in the pipeline.

silene commented:
Sure, and I indeed inferred that much from the commits. But that does not tell me why we should care. Despite being a heavy user of VM, I have never even encountered such a failure about `Meta`. So, this begs the question: Was the code in Native just dead? And so is now the code in Gen.

ppedrot commented:
No, I don't think that the native code for metas is dead. It is just that nowadays, it is very uncommon to encounter terms with metas. They only exist transiently in the unification engine and are bound to be removed eventually, but you may find them still. If you don't use VM compute over very dependent code that used stuff like `eexists` and friends, you're probably never going to experience the VM anomaly, but I am confident I can write some small example that exercises the code path.

silene commented:
> I am confident I can write some small example that exercises the code path.

That would be great, since I am pretty sure absolutely no user script has ever exercised it. Because otherwise, we would get report from people who have `native_compute` fall back to `vm_compute` (e.g., all the people who use Opam but do not install `coq-native`).

silene commented:
> I agree with this claim on a post 8.5 Coq version, but historically metas were everywhere.

This kind of reinforces my claim, since `native_compute` only ever existed in the post-8.5 world.

> It does not hurt to support the meta case in native compilation, still.

It does hurt! Code that cannot be exercised is just a long-term maintenance nightmare. We have no way to know whether this code is still correct, assuming it ever was. (And we are not talking about some random piece of OCaml code here, we are talking about code that calls `Obj.magic`.) This code should just be killed.

ppedrot commented:
I am fine with removing the support for metas in the native compiler, but I don't think this is directly related to this PR. It would be drowning a potentially important change in an ocean of trivial tweaks. If you want I can open another PR on top of this one. I'm not very concerned about the feasibility of meta removing, thanks to the miracle of pattern-matching that will be a couple of well-delimited cases.

ppedrot commented:
I consider this ready, see #16692 for the meta removal. Any assignee?

gares commented:
for what is worth, ttbomk ssr tactics don't generate terms with metas

ppedrot commented:
Well, there is the infamous function `Ssrmatching.evars_for_FO` that turns evars into metas for the sake of better-behaved unification.

gares commented:
yes, but then it tosses the result and does another unification wit evars. One day I'll rewrite a custom FO unif for that pass. It is just that back then using evarcon as a FO algorithm was broken, bacause it would also be FO on types (when checking instantiations).

ppedrot commented:
Ping for assignee. @silene maybe?

SkySkimmer commented on kernel/genlambda.ml:
```diff
@@ -0,0 +1,502 @@
+(* - Reducing [let] is the definition can be substituted *)
+(* - Transforming beta redex into [let] expression *)
+(* - Moving arguments under [let] *)
+(* Invariant: Terms in [subst] are already simplified and can be substituted                                            *)
```
```suggestion
(* Invariant: Terms in [subst] are already simplified and can be substituted *)
```

SkySkimmer reviewed (approved):
Looks OK

SkySkimmer commented:
merge tomorrow if no opposition

SkySkimmer commented:
@coqbot merge now
